lib/ocpp/v201/smart_charging.cpp

```diff
@@ -614,4 +625,126 @@ CompositeSchedule SmartChargingHandler::calculate_composite_schedule(
     return composite_schedule;
 }
 
+ChargingSchedule create_schedule_from_limit(const ConstantChargingLimit limit) {
+    return ChargingSchedule{
+        .id = 0,
+        .chargingRateUnit = limit.charging_rate_unit,
+        .chargingSchedulePeriod = {ChargingSchedulePeriod{
+            .startPeriod = 0,
+            .limit = limit.limit,
+        }},
+    };
+}
+
+std::optional<NotifyChargingLimitRequest>
+SmartChargingHandler::handle_external_limits_changed(const std::variant<ConstantChargingLimit, ChargingSchedule>& limit,
+                                                     double percentage_delta, ChargingLimitSourceEnum source,
+                                                     std::optional<int32_t> evse_id) const {
+    // K12.FR.04
+    if (source == ChargingLimitSourceEnum::CSO) {
+        // The spec does not define what we should do when the source
+        // given is CSO. Here we just throw.
+        throw std::invalid_argument("The source of an external limit should not be CSO.");
+    }
+
+    std::optional<NotifyChargingLimitRequest> request = {};
+
+    const auto& limit_change_cv = ControllerComponentVariables::LimitChangeSignificance;
+    const float limit_change_significance = this->device_model->get_value<double>(limit_change_cv);
+
+    const auto& notify_charging_limit_cv = ControllerComponentVariables::NotifyChargingLimitWithSchedules;
+    const std::optional<bool> notify_with_schedules =
+        this->device_model->get_optional_value<bool>(notify_charging_limit_cv);
+
+    if (percentage_delta > limit_change_significance) {
+        request = NotifyChargingLimitRequest{};
+        request->evseId = evse_id;
+        request->chargingLimit = {.chargingLimitSource = source};
+        if (notify_with_schedules.has_value() && notify_with_schedules.value()) {
+            if (const auto* limit_c = std::get_if<ConstantChargingLimit>(&limit)) {
+                request->chargingSchedule = {{create_schedule_from_limit(*limit_c)}};
+            } else if (const auto* limit_s = std::get_if<ChargingSchedule>(&limit)) {
+                request->chargingSchedule = {{*limit_s}};
+            }
+        }
```

Comment:
> The Charging Station SHALL inform the CSMS of the new **charging limit/schedule** imposed by the external system by sending a NotifyChargingLimitRequest.

We can be sent either a charging limit that's just a limit and the unit, or we can be sent a full charging schedule. Thus our methods need to support both. This method uses C++'s `variant` to support receiving either.

`ConstantChargingLimit` is a structured representation of the case where we receive only a limit and not a full schedule. In order to notify the CSMS via a `NotifyChargingLimitRequest`, we must represent whatever we receive as a charging schedule. So we have to translate the given charging limit into a schedule if we receive that instead of a schedule.